### PR TITLE
[WIP] Migrate to use null safety

### PR DESCRIPTION
**Pull request by raapperez**
Migrate all the code to support null safety

Next steps:
- [ ] Verify all the automatic fixes make sense
- [ ] Fix the example `test/next/nu_module_example.dart`
- [ ] Run the new version in the app

### CODE REVIEW

File: test/next/nu_module_example.dart [hunk 1]
```diff
@@ -47,11 +46,12 @@ class FriendRequestRouteExample
...
-    return Nuvigator(
-      router: null,
-      screenType: materialScreenType,
-    );
+      BuildContext context, NuRouteSettings<FriendRequestArgs?> settings) {
+    return Container();
+    // return Nuvigator(
+    //   router: null,
+    //   screenType: materialScreenType,
+    // );
```
Comment (raapperez) [inline test/next/nu_module_example.dart:54]: This example should be adjusted, `null` router is not acceptable anymore.

File: lib/src/legacy_nurouter.dart [hunk 2]
```diff
@@ -177,20 +180,20 @@ abstract class NuRouter implements INuRouter {
 
   /// From a deepLink (plus some option parameters) get a Route.
   @override
-  Route<T> getRoute<T>({
-    String deepLink,
-    Object parameters,
+  Route<T>? getRoute<T>({
+    String? deepLink,
```
Comment (leoiacovini) [inline lib/src/legacy_nurouter.dart:184]: Probably should be required

File: lib/src/legacy_nurouter.dart [hunk 3]
```diff
@@ -212,7 +215,7 @@ abstract class NuRouter implements INuRouter {
     if (isFromNative) {
       _addNativePopCallBack(route);
     }
-    return route;
+    return route as Route<T>?;
```
Comment (leoiacovini) [inline lib/src/legacy_nurouter.dart:218]: I have been noticed that type inference seems to be gotten worse, is this expected?

File: lib/src/next/v1/nu_router.dart [hunk 4]
```diff
@@ -139,18 +140,18 @@ abstract class NuRouter implements INuRouter {
 
   /// Override to true to call and register the routes only after the NuRouter initialization
   bool lazyRouteRegister = false;
-  List<NuRoute> _routes;
-  List<legacy.NuRouter> _legacyRouters;
-  NuvigatorState _nuvigator;
+  List<NuRoute>? _routes;
+  late List<legacy.NuRouter> _legacyRouters;
```
Comment (leoiacovini) [inline lib/src/next/v1/nu_router.dart:144]: Why the late here vs making nullable elsewhere?

File: lib/src/next/v1/nu_router.dart [hunk 5]
```diff
@@ -252,34 +253,34 @@ abstract class NuRouter implements INuRouter {
     }
   }
 
-  ScreenRoute<R> _getScreenRoute<R>(String deepLink,
-      {Map<String, dynamic> parameters}) {
-    for (final route in routes) {
+  ScreenRoute<R>? _getScreenRoute<R>(String? deepLink,
```
Comment (leoiacovini) [inline lib/src/next/v1/nu_router.dart:256]: I think deepLink should not be allowed to be null

File: lib/src/next/v1/nu_router.dart [hunk 6]
```diff
@@ -252,34 +253,34 @@ abstract class NuRouter implements INuRouter {
...
     }
     return null;
   }
 
   @override
-  Route<R> getRoute<R>({
-    String deepLink,
-    Object parameters,
+  Route<R>? getRoute<R>({
+    String? deepLink,
```
Comment (leoiacovini) [inline lib/src/next/v1/nu_router.dart:270]: required probably

File: lib/src/nuvigator.dart [hunk 7]
```diff
@@ -26,35 +26,35 @@ NuvigatorState _tryToFindNuvigatorForRouter<T extends INuRouter>(
 }
 
 class NuvigatorStateTracker extends NavigatorObserver {
-  final List<Route> stack = [];
+  final List<Route?> stack = [];
```
Comment (leoiacovini) [inline lib/src/nuvigator.dart:29]: I think `Route` shouldn't be nullable

File: lib/src/nuvigator.dart [hunk 8]
```diff
@@ -66,36 +66,35 @@ abstract class INuRouter {
...
 
   @deprecated
-  T getRouter<T extends INuRouter>();
+  T? getRouter<T extends INuRouter>();
 
-  Route<T> getRoute<T>({
-    String deepLink,
-    Object parameters,
+  Route<T>? getRoute<T>({
+    String? deepLink,
```
Comment (leoiacovini) [inline lib/src/nuvigator.dart:75]: Probably should be required non-nullable

File: lib/src/screen_route.dart [hunk 9]
```diff
@@ -9,8 +9,8 @@ import 'typings.dart';
 class RouteDef {
   RouteDef(this.routeName, {this.deepLink});
 
-  final String routeName;
-  final String deepLink;
+  final String? routeName;
```
Comment (leoiacovini) [inline lib/src/screen_route.dart:12]: Probably shouldn't be nullable

**Review by leoiacovini — COMMENTED** [review]
Left come comments so we can check later in more depth

File: lib/src/next/v1/nu_router.dart (hunk 4 above)
Comment (raapperez) [reply lib/src/next/v1/nu_router.dart:144]: Late is used when it's not set yet but you can guarantee it's not null when it's used.

File: lib/src/legacy_nurouter.dart (hunk 3 above)
Comment (raapperez) [reply lib/src/legacy_nurouter.dart:218]: In this case it can't infere Route<dynamic> as Route<T>?

File: lib/src/legacy_nurouter.dart [hunk 10]
```diff
@@ -107,7 +105,7 @@ abstract class NuRouter implements INuRouter {
   }
 
   /// Deprecated: Prefer using getRoute
-  RouteEntry getRouteEntryForDeepLink(String deepLink) {
+  RouteEntry? getRouteEntryForDeepLink(String? deepLink) {
```
Comment (leoiacovini) [inline lib/src/legacy_nurouter.dart:108]: I think `deepLink` is required

**Review by leoiacovini — COMMENTED** [review]
I think overall it's ok :)

**Pull request closed without merging**